### PR TITLE
Warning on eval module

### LINKED ISSUE #20986: When the model is in an eval state before calling trainer.fit it should be moved to train state
### Bug description

If somehow the model is set to the `eval` mode (calling `model.eval` or model.train(False)`) before calling teh trainer.fit supposed to trained the model, the model is staying in the `eval` mode, thus not applying some tricks interesting for training.

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
import torch
from torch.utils.data import Dataset, DataLoader
import pytorch_lightning as pl

# Dummy dataset
class DummyDataset(Dataset):
    def __init__(self, length=5):
        self.length = length
        self.x = torch.randn(length, 10)
        self.y = torch.randn(length, 1)

    def __len__(self):
        return self.length

    def __getitem__(self, idx):
        return self.x[idx], self.y[idx]

# Simple LightningModule
class SimpleModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 1)
        self.loss_fn = torch.nn.MSELoss()

    def forward(self, x):
        return self.linear(x)

    def training_step(self, batch, batch_idx):
        print(f"is the model in train state: {self.training}")
        x, y = batch
        y_hat = self(x)
        loss = self.loss_fn(y_hat, y)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)

# Instantiate dataset and dataloader
dataset = DummyDataset()
dataloader = DataLoader(dataset, batch_size=16)

# Instantiate model and trainer
model = SimpleModel()
model.eval()
trainer = pl.Trainer(max_epochs=1, logger=False, enable_checkpointing=False)

# Train
trainer.fit(model, dataloader)
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 4060 Ti
                - NVIDIA GeForce GTX 1080 Ti
        - available:         True
        - version:           12.6
* Lightning:
        - lightning:         2.5.0
        - lightning-utilities: 0.14.3
        - pytorch-lightning: 2.5.1.post0
        - torch:             2.7.1
        - torchmetrics:      1.7.3
* Packages:
        - aiohappyeyeballs:  2.6.1
        - aiohttp:           3.12.13
        - aiosignal:         1.3.2
        - alembic:           1.16.2
        - annotated-types:   0.7.0
        - antlr4-python3-runtime: 4.9.3
        - anyio:             4.9.0
        - async-timeout:     5.0.1
        - attrs:             25.3.0
        - autocommand:       2.2.2
        - backports.tarfile: 1.2.0
        - blinker:           1.9.0
        - cachetools:        5.5.2
        - certifi:           2025.6.15
        - charset-normalizer: 3.4.2
        - click:             8.2.1
        - cloudpickle:       3.1.1
        - contourpy:         1.3.2
        - cycler:            0.12.1
        - databricks-sdk:    0.57.0
        - docker:            7.1.0
        - exceptiongroup:    1.3.0
        - fastapi:           0.115.13
        - filelock:          3.18.0
        - flask:             3.1.1
        - fonttools:         4.58.4
        - frozenlist:        1.7.0
        - fsspec:            2025.5.1
        - gitdb:             4.0.12
        - gitpython:         3.1.44
        - google-auth:       2.40.3
        - graphene:          3.4.3
        - graphql-core:      3.2.6
        - graphql-relay:     3.2.0
        - greenlet:          3.2.3
        - gunicorn:          23.0.0
        - h11:               0.16.0
        - idna:              3.10
        - importlib-metadata: 8.7.0
        - inflect:           7.3.1
        - itsdangerous:      2.2.0
        - jaraco.collections: 5.1.0
        - jaraco.context:    5.3.0
        - jaraco.functools:  4.0.1
        - jaraco.text:       3.12.1
        - jinja2:            3.1.6
        - joblib:            1.5.1
        - kiwisolver:        1.4.8
        - lightning:         2.5.0
        - lightning-utilities: 0.14.3
        - mako:              1.3.10
        - markupsafe:        3.0.2
        - matplotlib:        3.10.3
        - mlflow:            3.1.0
        - mlflow-skinny:     3.1.0
        - more-itertools:    10.3.0
        - mpmath:            1.3.0
        - multidict:         6.5.0
        - networkx:          3.4.2
        - numpy:             2.2.6
        - nvidia-cublas-cu12: 12.6.4.1
        - nvidia-cuda-cupti-cu12: 12.6.80
        - nvidia-cuda-nvrtc-cu12: 12.6.77
        - nvidia-cuda-runtime-cu12: 12.6.77
        - nvidia-cudnn-cu12: 9.5.1.17
        - nvidia-cufft-cu12: 11.3.0.4
        - nvidia-cufile-cu12: 1.11.1.6
        - nvidia-curand-cu12: 10.3.7.77
        - nvidia-cusolver-cu12: 11.7.1.2
        - nvidia-cusparse-cu12: 12.5.4.2
        - nvidia-cusparselt-cu12: 0.6.3
        - nvidia-nccl-cu12:  2.26.2
        - nvidia-nvjitlink-cu12: 12.6.85
        - nvidia-nvtx-cu12:  12.6.77
        - omegaconf:         2.3.0
        - opentelemetry-api: 1.34.1
        - opentelemetry-sdk: 1.34.1
        - opentelemetry-semantic-conventions: 0.55b1
        - packaging:         24.2
        - pandas:            2.3.0
        - pillow:            11.2.1
        - platformdirs:      4.2.2
        - propcache:         0.3.2
        - protobuf:          6.31.1
        - pyarrow:           20.0.0
        - pyasn1:            0.6.1
        - pyasn1-modules:    0.4.2
        - pydantic:          2.11.7
        - pydantic-core:     2.33.2
        - pyparsing:         3.2.3
        - python-dateutil:   2.9.0.post0
        - pytorch-lightning: 2.5.1.post0
        - pytz:              2025.2
        - pyyaml:            6.0.2
        - requests:          2.32.4
        - rsa:               4.9.1
        - scikit-learn:      1.7.0
        - scipy:             1.15.3
        - setuptools:        80.9.0
        - six:               1.17.0
        - smmap:             5.0.2
        - sniffio:           1.3.1
        - sqlalchemy:        2.0.41
        - sqlparse:          0.5.3
        - starlette:         0.46.2
        - sympy:             1.14.0
        - tabulate:          0.9.0
        - threadpoolctl:     3.6.0
        - tomli:             2.2.1
        - torch:             2.7.1
        - torchmetrics:      1.7.3
        - tqdm:              4.67.1
        - triton:            3.3.1
        - typeguard:         4.3.0
        - typing-extensions: 4.14.0
        - typing-inspection: 0.4.1
        - tzdata:            2025.2
        - urllib3:           2.5.0
        - uvicorn:           0.34.3
        - werkzeug:          3.1.3
        - wheel:             0.45.1
        - yarl:              1.20.1
        - zipp:              3.23.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.10.14
        - release:           6.8.0-59-generic
        - version:           #61~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 15 17:03:15 UTC 2

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #20986
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/20216
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/20573
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/20177

Adds a simple warning if any module when `fit` starts are in eval mode. There are cases where you intentionally want modules in eval mode even during `fit`, however most users want things to be in train mode. This warning should hopefully help users identify this can be a potential problem.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/loops/fit_loop.py [hunk 1]
```diff
@@ -515,6 +518,19 @@ def on_load_checkpoint(self, state_dict: dict) -> None:
...
 
+    def _warn_if_modules_in_eval_mode(self) -> None:
+        """Warn if any modules are in eval mode at the start of training."""
+        model = self.trainer.lightning_module
+        eval_modules = [name for name, module in model.named_modules() if not module.training]
+
+        if eval_modules:
+            rank_zero_warn(
+                f"Found {len(eval_modules)} module(s) in eval mode at the start of training."
+                f" This may lead to unexpected behavior during training. If this is intentional,"
```
Comment (deependujha) [inline src/lightning/pytorch/loops/fit_loop.py]:
```suggestion
                " This may lead to unexpected behavior during training. If this is intentional,"
```
